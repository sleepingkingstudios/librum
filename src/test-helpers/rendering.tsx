import * as React from 'react';
import { Provider as ReduxProvider } from 'react-redux';
import {
  MemoryRouter,
  Routes,
  Route
} from "react-router-dom";
import {
  RenderOptions,
  RenderResult,
  render,
} from '@testing-library/react';

import {
  createStore,
  storeType,
} from './store';
import {
  ThemeContext,
  defaultTheme,
} from '@themes/context';
import type { Theme } from '@themes';

interface customRenderOptions extends RenderOptions {
  initialEntries?: Array<string>;
  router?: true | withRouterOptions;
  store?: (storeType | true);
  theme?: (Theme | true);
}

interface withRouterOptions {
  initialEntries?: Array<string>;
  path?: string;
}

export const withRouter = (
  component: React.ReactElement,
  {
    initialEntries = ['/'],
    path,
  }: withRouterOptions,
): React.ReactElement => {
  let wrapped = component;

  if (path) {
    wrapped = (
      <Routes>
        <Route path={path} element={wrapped} />
      </Routes>
    );
  }

  return (
    <MemoryRouter initialEntries={initialEntries}>
      { wrapped }
    </MemoryRouter>
  );
};

export const withStore = (
  component: React.ReactElement,
  { store }: { store: (storeType | true) }
): React.ReactElement => {
  let configuredStore: storeType;

  if (store === true) {
    configuredStore = createStore().store;
  } else {
    configuredStore = store;
  }

  return (
    <ReduxProvider store={configuredStore}>
      { component }
    </ReduxProvider>
  );
};

export const withTheme = (
  component: React.ReactElement,
  { theme }: { theme: (Theme | true) }
): React.ReactElement => {
  let configuredTheme: Theme;

  if (theme === true) {
    configuredTheme = defaultTheme;
  } else {
    configuredTheme = theme;
  }

  return (
    <ThemeContext.Provider value={configuredTheme}>
      { component }
    </ThemeContext.Provider>
  );
};

const customRender = (
  component: React.ReactElement,
  options?: customRenderOptions,
  strictMode = true,
): RenderResult => {
  let wrapped: React.ReactElement = component;

  if (strictMode) {
    wrapped = (<React.StrictMode>{ wrapped }</React.StrictMode>);
  }

  if (!options) { return render(wrapped); }

  if ('router' in options && options.router) {
    if (options.router === true) {
      wrapped = withRouter(wrapped, { initialEntries: options.initialEntries });
    } else {
      wrapped = withRouter(
        wrapped,
        {
          initialEntries: options.initialEntries,
          ...options.router,
        },
      );
    }
  }

  if ('store' in options) {
    wrapped = withStore(wrapped, { store: options.store });
  }

  if ('theme' in options) {
    wrapped = withTheme(wrapped, { theme: options.theme });
  }

  return render(wrapped, options);
};

export { customRender as render };
