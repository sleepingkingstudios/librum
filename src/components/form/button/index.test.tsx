import * as React from 'react';

import '@testing-library/jest-dom';
import { render } from '@test-helpers/rendering';

import { FormButton } from './index';

describe('<FormButton />', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(
      <FormButton>Button</FormButton>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('with className: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton className="custom-button">Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with children: elements', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton className="custom-button">
          <span className="custom-icon" />
          Button
        </FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with cols: value', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton cols={3}>Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with disabled: true', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton disabled>Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: button', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton type="button">Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: reset', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton type="reset">Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('with type: submit', () => {
    it('should match the snapshot', () => {
      const { asFragment } = render(
        <FormButton type="submit">Button</FormButton>,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
