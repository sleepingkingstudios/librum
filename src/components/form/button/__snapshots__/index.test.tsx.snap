// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormButton /> should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button"
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with children: elements should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button custom-button"
    type="button"
  >
    <span
      class="custom-icon"
    />
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with className: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button custom-button"
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with cols: value should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button col-span-3"
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with disabled: true should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button"
    disabled=""
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with type: button should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button"
    type="button"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with type: reset should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button"
    type="reset"
  >
    Button
  </button>
</DocumentFragment>
`;

exports[`<FormButton /> with type: submit should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="form-button"
    type="submit"
  >
    Button
  </button>
</DocumentFragment>
`;
